Test stability: FlowRunnerTest2

juhoautio opened the pull request:
**Let's have https://github.com/azkaban/azkaban/pull/1159 before this, ok?**

Calling `ExecutableFlowBase#isFlowFinished` from the test thread wasn't a good idea.

In rare cases it failed with this exception (see https://github.com/azkaban/azkaban/pull/1157#issuecomment-306055490):

```
azkaban.execapp.FlowRunnerTest2 > testCancel FAILED
java.lang.NullPointerException
at azkaban.executor.ExecutableFlowBase.isFlowFinished(ExecutableFlowBase.java:398)
at azkaban.execapp.FlowRunnerTestBase.lambda$assertThreadShutDown$0(FlowRunnerTestBase.java:32)
at azkaban.execapp.FlowRunnerTestBase$$Lambda$9/1157486615.apply(Unknown Source)
at azkaban.execapp.FlowRunnerTestBase.waitFlowRunner(FlowRunnerTestBase.java:42)
at azkaban.execapp.FlowRunnerTestBase.assertThreadShutDown(FlowRunnerTestBase.java:31)
at azkaban.execapp.FlowRunnerTest2.testCancel(FlowRunnerTest2.java:729)
```

Extract from `ExecutableFlowBase` (this method is actually deleted in this PR):

```java
  /**
   * Only returns true if the status of all finished nodes is true.
   */
  public boolean isFlowFinished() {
    for (final String end : getEndNodes()) {
      final ExecutableNode node = getExecutableNode(end);
      // THIS IS WHERE THE NPE HAPPENED, so node was null
      if (!Status.isStatusFinished(node.getStatus())) {
        return false;
      }
    }

    return true;
  }
```

It must be that the NPE happened because this method was being called from the test main thread, and there is no synchronization on the "end nodes". So it was somehow possible that the test got a name for an end node but then couldn't get the ExecutableNode to match it.

I used `isFlowFinished()` (it existed previously, I didn't add it) in the test because it seemed like a handy way to check that, but seems that it didn't work reliably. And now I realized that this method is not being used at all otherwise, so it can be just removed.

Now checking flow completion in a thread-safe way.

Also delete unused methods in `ExecutableFlowBase`:
- `isFlowFinished`
- `findNextJobsToRun`

**Make `ExecutableNode.status` volatile**
- This gives guarantees especially for unit tests that need to check for statuses in a multi-threaded environment.
- But also `FlowRunner` thread vs. `JobRunner` threads benefit from this, although I haven't been able to prove that there would be any critical issues there.

HappyRay commented:
#1155

juhoautio commented:
Regardless of the build failure, this change is for the better, because it narrows down the ways of checking statuses concurrently.

HappyRay commented:
Could you share more details on the problem and how the fix works?

HappyRay commented:
It seems to me that both the production code and tests are complex given all the trouble we have. Can you think of ways to simplfy? e.g. can we do without mutliple threads? I understand this may require a large refactoring...

juhoautio commented:
> Could you share more details on the problem and how the fix works?

Added to PR description.

juhoautio commented:
> Can you think of ways to simplfy? e.g. can we do without mutliple threads?

Sure there are different ways to simplify and make this more robust. Maybe the simplest way would be to keep the threads as they are but use some concurrent utilities to make things easier to manage.

HappyRay commented:
Please see another new failure.

Thanks for the explanation.
Why is the new code thread safe?

juhoautio commented:
> Why is the new code thread safe?

Because it just checks the flow status, not statuses of every job in the flow. And this is enough to check if the flow has finished.

juhoautio commented:
Added commit: "Make ExecutableNode.status volatile". Also added to PR description. After that change I just ran `FlowRunnerTest2` 1000 times in a row locally and didn't get a single test failure.

juhoautio commented:
Same thing with `FlowRunnerTest` – ran 1000 times locally without failures (after making `ExecutableNode.status` volatile).

juhoautio commented:
Now that I merged from master there are some of those `@Ignore`s added. Once this has been merged I can create a new PR to remove the `@Ignore`s.

juhoautio commented:
Made a separate PR for that volatile status field: https://github.com/azkaban/azkaban/pull/1159

HappyRay commented:
I see. 
The getStatus method is only thread-safe after the status field is made volatile. 

LGTM.
